### ISSUE TITLE
Comments wrongly converted in lists

### ISSUE THREAD

**Issue by duncombe** [russross/blackfriday#136]
I am using [Hugo](github.com/spf13/hugo), and encountered the bug detailed below. I am told blackfriday is the markdown parser used by Hugo, so I am reporting the bug here too.

Within a list (e.g., `<DL>` `</DL>`) in a markdown document, Hugo converts comments (`<!-- -->`) to `&lt;!&ndash; &ndash;&gt;`. Outside the list scope the comments are treated correctly and left as comments (`<!-- -->`). This is a bug.

I first observed this bug in definition lists, but it turns out all lists are affected, including markdown lists demarcated with `-` or `*`, not just HTML tagged lists, with the comments incorrectly rendered, turning them into displayed characters. The bug can be circumvented by separating the comments from the list using copious amounts of blank lines, but these blank lines then appear in the HTML affecting the layout of the list.

Refers  to spf13/hugo#729.

**Comment by anthonyfok** [comment]
It is not just lists... Paragraphs too!  When I tested with https://github.com/russross/blackfriday-tool:

    Hello 

gives

> <p>Hello &lt;!&ndash; there &ndash;&gt;</p>

    <p>Hello &lt;!&ndash; there &ndash;&gt;</p>

This is somehow related to `smartDash()` and/or `smartDashLatex()` in `smartypants.go`, as well as `htmlComment()` in `block.go`.

**Comment by yacoob** [comment]
Another case where it misbehaves, quoted from a [Hugo bug](https://github.com/spf13/hugo/issues/1080)

Source code like this:

```markdown
blahblah
<!--- foo -->
rhubarb
```

becomes

```html
blahblah
-&lt;!&ndash; foo &ndash;&gt;
rhubarb
```

If you separate comment in the source file by extra empty lines, the conversion is inhibited *but* extra <p> are added:

```markdown
blahblah
<!--- foo -->
rhubarb
```

becomes

```html
blahblah
</p>
<!-- foo -->
<p>
rhubarb
```

This is annoying :(

**Comment by thatguystone** [comment]
This is probably the wrong way to fix it (it's a bit hacky), but it gets the job done: 9e5f8e534b83b9a15611f5af53600b9f2727aeae

**Comment by bep** [comment]
Any plans to fix this?

**Comment by rtfb** [comment]
@anthonyfok, am I right thinking that #190 is a workaround for this problem?

**Comment by rtfb** [comment]
I have just pushed a fix: https://github.com/russross/blackfriday/commit/5d3d5c198ebf779798f95e73e990ecc271db3735

Please take a look (and a test run if possible). More obscure test cases also welcome, while we're at it.

**Comment by anthonyfok** [comment]
Hi @rtfb, thank you for fixing this issue #190 too!

Sorry for my tardy reply as I missed your message earlier.  Here is my reply:

> No, #190 is not a suggested fix to this issue, or at least not that I intended.  :-)  I have forgotten about this issue when I submitted #190, which deals with the appearance of smart dashes *outside* of comments.

But since you have already kindly both fixed #136 and merged #190, my after-the-fact reply is irrelevant now.  ;-)

Cheers,
Anthony
